src-ui.v3/src/API.hs

```diff
@@ -358,3 +282,283 @@ labelModCml = uncap        . dropWhile (not . C.isUpper)
+  , _putV2PackageTags :: Dynamic t (Either Text TagN) -> Dynamic t (Either Text PkgN) -> Event t () -> m (Event t (ReqResult () NoContent))
+  , _deleteV2PackageTags :: Dynamic t (Either Text TagN) -> Dynamic t (Either Text PkgN) -> Event t () -> m (Event t (ReqResult () NoContent))  
+  , _getV2UnitInfo :: Dynamic t (Either Text UUID) -> Event t () -> m (Event t (ReqResult () UnitIdInfo))
+  , _getV2User :: Dynamic t (Either Text UserName) -> Event t () -> m (Event t (ReqResult () UserPkgs))
+  , _getV2Workers :: Event t () -> m (Event t (ReqResult () (Vector WorkerRow)))
+  , _getV2WorkersPkg :: Dynamic t (Either Text PkgN) -> Event t () -> m (Event t (ReqResult () (Vector WorkerRow)))
+  }
+makeLenses ''ClientFuns
+
+getClient :: forall t m. (SupportsServantReflex t m) 
+  => ClientFuns t m 
+getClient = mkClientFuns'' burlNew
+  where 
+    mkClientFuns'' bp = ClientFuns { .. }
+      where
+        (      _getV2IdxStates 
+          :<|> _getV2Info
+          :<|> _getV2Packages
+          :<|> _getV2PackagesHistory
+          :<|> _getV2PackageHistory
+          :<|> _getV2PackageReports
+          :<|> _getV2PackageReportSummary
+          :<|> _getV2PackageReportDetail
+          :<|> _getV2PackageTags
+          :<|> _getV2Queue
+          :<|> _getV2QueuePkg
+          :<|> _putV2Queue
+          :<|> _getV2TagsWithPackage
+          :<|> _getV2TagsWithoutPackage
+          :<|> _putV2PackageTags
+          :<|> _deleteV2PackageTags
+          :<|> _getV2UnitInfo
+          :<|> _getV2User
+          :<|> _getV2Workers
+          :<|> _getV2WorkersPkg ) = (clientWithOpts (Proxy :: Proxy API) Proxy (Proxy :: Proxy ()) (constDyn bp) tweakRequest) :: Client t m API ()
+
+type ValidationErrors = Map Text [Text]
+type ClientRes t a = (Event t a, Event t ClientError, Dynamic t Bool)
+  
+data ClientError
+  =   Forbidden
+  | NotFound
+  | Unauthorised
+  | FailedValidation (Maybe (ErrorBody ValidationErrors))
+  | OtherError Word Text
+  deriving (Show)
+  
+data ErrorBody errors = ErrorBody
+  { message :: Text
+  , errors  :: Maybe errors
+  } deriving (Generic, Show)
+        
+deriving instance ToJSON errors   => ToJSON (ErrorBody errors)
+deriving instance FromJSON errors => FromJSON (ErrorBody errors)
+
+wireClientRes
+  :: (Reflex t, MonadHold t m)
+  => Event t b
+  -> Event t (ReqResult () a)
+  -> m (ClientRes t a)
+wireClientRes evSubmit resE = do
+  let evSuccess = fmapMaybe reqSuccess resE
+  let evError   = fmapMaybe reqClientError resE
+  submittingDyn <- holdDyn False $ leftmost [True <$ evSubmit, False <$ evError, False <$ evSuccess]
+  pure (evSuccess, evError, submittingDyn)
+
+reqClientError :: ReqResult tag a -> Maybe ClientError
+reqClientError (ResponseFailure _ msg xhrR) = Just $ case view xhrResponse_status xhrR of
+  401 -> Unauthorised
+  403 -> Forbidden
+  404 -> NotFound
+  422 -> FailedValidation (xhrR ^? xhrResponse_responseText . _Just . to fromStrict . to encodeUtf8 . to decode . _Just)
```

Comment:
This part here is real world specific. You will need to base this on whether the backend returns a response body on specific error codes. There's no HTTP standard here and maybe the backend actually never returns any meaningful error except the status code. 

I'd have to dig into the backend code for that. I can do that if you'd like. :slightly_smiling_face:

Reply:
I have look into the backend and it only shows basic throw created here: https://github.com/haskell-CI/hackage-matrix-builder/blob/master/src-exe/Controller/WebSvc.hs#L593 (e.g 401, 304, 404) do you mean that the 422 and 403 is not necessary?

Reply:
Yeah, looks like 403 and 422 are not necessary. Which means that error body can definitely die.